### PR TITLE
Add a .wasm cache in wasm-c-api

### PR DESCRIPTION
**Pull request by lum1n0us**
Use sha256 to hash binary file content. If the incoming .wasm is cached before, wasm_module_new() simply returns the existed one.

### CODE REVIEW

File: core/iwasm/common/wasm_c_api.c [hunk 1]
```diff
@@ -17,6 +18,10 @@
 #endif /*WASM_ENABLE_JIT != 0 && WASM_ENABLE_LAZY_JIT == 0*/
 #endif /*WASM_ENABLE_AOT != 0*/
 
+#if WASM_ENABLE_WASM_CACHE != 0
+#include "openssl/sha.h"
```
Comment (wenyongh) [inline core/iwasm/common/wasm_c_api.c:22]: Had better `<openssl/sha.h>`?

**Comment by yamt** [comment]
a question: is it safe to share a module at all?
to me, wasm_instance_new seems to modify the module for linking.

File: core/iwasm/common/wasm_c_api.c [hunk 2]
```diff
@@ -1911,16 +1919,75 @@ module_to_module_ext(wasm_module_t *module)
...
+    return NULL;
+}
+
+static wasm_module_ex_t *
+try_reuse_loaded_module(wasm_store_t *store, char *binary_hash)
+{
+    wasm_module_ex_t *cached = NULL;
+    wasm_module_ex_t *ret = NULL;
+
+    cached = check_loaded_module(&singleton_engine->modules, binary_hash);
```
[the remark this answers is not in the document]
Comment (lum1n0us) [reply core/iwasm/common/wasm_c_api.c:1959]: I am thinking about a *lazy protection* here. It is based on *wasm_module_delete_internal* just resetting fields and *wasm_engine_delete_internal* actually releasing resource of *wasm_module_t*. So, L1957 `if (!cached->ref_count)` will prevent any further action on a deleted module.

**Review by wenyongh — APPROVED** [review]
LGTM

**Comment by yamt** [comment]
> @yamt Would you be more specific? I remembered that loader will change Wasm module related data structures. But not during _instantiation_.

eg. interp_link_func modifies module_interp->import_functions[x].

**Comment by lum1n0us** [comment]
I see the point. There is a gap about linking between *wasm-c-api* and *WAMR_BUILD_MULTI_MODULE* 
In *wasm-c-api*, the linking action is during instantiate. But in *WAMR_BUILD_MULTI_MODULE*, it happens in loading phase.

**Comment by yamt** [comment]
> I see the point. There is a gap about linking between _wasm-c-api_ and _WAMR_BUILD_MULTI_MODULE_ In _wasm-c-api_, the linking action is during instantiate. But in _WAMR_BUILD_MULTI_MODULE_, it happens in loading phase.

yea. and it's actually wasm-c-api this PR is about.

**Pull request merged**